### PR TITLE
Add a temporary PAL for ClientWebSocket.

### PR DESCRIPTION
**Pull request by pgavlin**
On Windows, the PAL delegates to WinHttpWebSocket; on *nix, all methods
throw PlatformNotSupportedException.

### CODE REVIEW

**Comment by pgavlin** [comment]
@CIPop @davidsh @roncain @stephentoub PTAL

**Comment by pgavlin** [comment]
FWIW, I'm not sure that this will be the final shape of the PAL. This is primarily to unblock @roncain and the rest of WCF.

**Comment by CIPop** [comment]
+ @mconnew since this might cause merge issues with the internal changes.

File: src/System.Net.WebSockets.Client/src/System.Net.WebSockets.Client.csproj [hunk 1]
```diff
@@ -10,17 +10,42 @@
...
+    </Compile>
+    <Compile Include="$(CommonPath)\System\Net\UriScheme.cs">
+      <Link>Common\System\Net\UriScheme.cs</Link>
+    </Compile>
+    <Compile Include="$(CommonPath)\System\Net\WebSockets\WebSocketValidate.cs">
+      <Link>Common\System\Net\WebSockets\WebSocketValidate.cs</Link>
+    </Compile>
+  </ItemGroup>
+
+  <ItemGroup Condition=" '$(TargetsWindows)' == 'true' ">
```
Comment (mconnew) [inline src/System.Net.WebSockets.Client/src/System.Net.WebSockets.Client.csproj]: I've just checked in via TFS some changes which will need to be accommodated here. There is now a ClientWebSocket.Windows.cs file. There will need to be a matching one for Unix which throws in the appropriate places.

File: src/System.Net.WebSockets.Client/src/System/Net/WebSockets/ClientWebSocket.cs [hunk 2]
```diff
@@ -156,7 +156,7 @@ public Task ConnectAsync(Uri uri, CancellationToken cancellationToken)
 
         private Task ConnectAsyncCore(Uri uri, CancellationToken cancellationToken)
         {
-            _innerWebSocket = new WinHttpWebSocket();
+            _innerWebSocket = WebSocketHandle.Create();
```
Comment (mconnew) [inline src/System.Net.WebSockets.Client/src/System/Net/WebSockets/ClientWebSocket.cs]: This isn't a sufficient abstraction for handling multiple platforms as the exception handling is insufficient for NetNative. The ConnectAsyncCore implementation needs to be pushed into platform specific code.

Comment (pgavlin) [inline src/System.Net.WebSockets.Client/src/System/Net/WebSockets/ClientWebSocket.cs]: Noted, but .NET Native-related changes are out of scope for this PR: the goal here is simply to create an implementation of ClientWebSocket for *nix platforms that does nothing but throw PNSE.

Comment (mconnew) [inline src/System.Net.WebSockets.Client/src/System/Net/WebSockets/ClientWebSocket.cs]: This PR will conflict with some code already checked in via TFS and is waiting on the existing tfs-git sync PR to be merged before it shows up. So the issue needs to be addressed whether it was in scope for this PR or not.

Comment (pgavlin) [inline src/System.Net.WebSockets.Client/src/System/Net/WebSockets/ClientWebSocket.cs]: Ah, that's clearer. :) I'm happy to wait for the code to flow in from TFS and adjust this change accordingly as long as @roncain is OK with waiting. @roncain: in the meantime, you could pull down https://github.com/pgavlin/corefx/tree/WebSocketClientPNSE and use the S.N.WebSockets.Client.dll out of the build.

(/cc @joshfree just as a heads-up)

**Comment by davidsh** [comment]
@dotnet-bot test this please

**Comment by pgavlin** [comment]
@davidsh @mconnew PTAL: I've updated this to account for the changes that came in with the merge.

**Comment by davidsh** [comment]
LGTM

File: src/System.Net.WebSockets.Client/src/System/Net/WebSockets/ClientWebSocket.cs [hunk 3]
```diff
@@ -146,6 +144,31 @@ public Task ConnectAsync(Uri uri, CancellationToken cancellationToken)
...
+
+                await _innerWebSocket.ConnectAsyncCore(uri, cancellationToken, _options);
+            }
+            catch (Exception ex)
+            {
+                if (Logging.On)
+                {
+                    Logging.Exception(Logging.WebSockets, this, "ConnectAsync", ex);
+                }
+                throw;
```
Comment (stephentoub) [inline src/System.Net.WebSockets.Client/src/System/Net/WebSockets/ClientWebSocket.cs]: Do we need to be resetting any state here, e.g. Dispose'ing and null'ing out _innerWebSocket, changing _state back to Connecting, etc.?

Comment (stephentoub) [inline src/System.Net.WebSockets.Client/src/System/Net/WebSockets/ClientWebSocket.cs]: Ah, after reading further, I see that you only moved this logic, so this question isn't specific to this PR.

File: src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Win32.cs [hunk 4]
```diff
@@ -25,22 +40,17 @@ public sealed partial class ClientWebSocket : WebSocket
...
             {
-                // Change internal state to 'connected' to enable the other methods
-                if ((InternalState)Interlocked.CompareExchange(ref _state, (int)InternalState.Connected, (int)InternalState.Connecting) != InternalState.Connecting)
-                {
-                    // Aborted/Disposed during connect.
-                    throw new ObjectDisposedException(GetType().FullName);
-                }
-
-                return winHttpWebSocket.ConnectAsync(uri, cancellationToken, _options);
+                await _webSocket.ConnectAsync(uri, cancellationToken, options);
```
Comment (stephentoub) [inline src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Win32.cs]: Should add .ConfigureAwait(false) to the end.

File: src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.WinRT.cs [hunk 5]
```diff
@@ -1,31 +1,43 @@
...
             {
-                // Change internal state to 'connected' to enable the other methods
-                if ((InternalState)Interlocked.CompareExchange(ref _state, (int)InternalState.Connected, (int)InternalState.Connecting) != InternalState.Connecting)
-                {
-                    // Aborted/Disposed during connect.
-                    throw new ObjectDisposedException(GetType().FullName);
-                }
-
-                await rtWebSocket.ConnectAsync(uri, cancellationToken, Options);
+                await rtWebSocket.ConnectAsync(uri, cancellationToken, options);
```
Comment (stephentoub) [inline src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.WinRT.cs]: Ditto on ConfigureAwait(false)

**Comment by stephentoub** [comment]
A few minor comments, otherwise LGTM.

File: src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Unix.cs [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+            }
+        }
+        #endregion
+
+        public static WebSocketHandle Create()
+        {
+            return default(WebSocketHandle);
+        }
+
+        public static void CheckPlatformSupport()
```
Comment (vijaykota) [inline src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Unix.cs]: Doubt: The semantics for Windows seem to be to throw if there is no support

**Comment by vijaykota** [comment]
LGTM. Thanks for fixing, @pgavlin
